src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxDocumentConverter.cs

```diff
@@ -0,0 +1,252 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Globalization;
+using Corax;
+using Raven.Client;
+using Raven.Server.Documents.Indexes.Persistence.Corax.WriterScopes;
+using Raven.Server.Json;
+using Raven.Server.Utils;
+using Sparrow.Extensions;
+using Sparrow.Json;
+using Sparrow.Json.Parsing;
+using Sparrow.Server;
+using Sparrow.Threading;
+using Voron;
+using Encoding = System.Text.Encoding;
+
+namespace Raven.Server.Documents.Indexes.Persistence.Corax
+{
+    public sealed class CoraxDocumentConverter : ConverterBase
+    {
+        private readonly ByteStringContext _allocator;
+        private readonly Dictionary<Slice, int> _knownFields;
+        private static readonly byte[] _trueLiteral = Encoding.UTF8.GetBytes("true"); 
+        private static readonly byte[] _falseLiteral = Encoding.UTF8.GetBytes("false"); 
+
+        public CoraxDocumentConverter(
+                Index index, 
+                bool indexImplicitNull = false, 
+                bool indexEmptyEntries = true, 
+                string keyFieldName = null, 
+                bool storeValue = false, 
+                string storeValueFieldName = Constants.Documents.Indexing.Fields.ReduceKeyValueFieldName) :
+            base(index, storeValue)
+        {
+            _allocator = new ByteStringContext(SharedMultipleUseFlag.None);
+            _knownFields = GetKnownFields();
+        }
+
+        public Dictionary<Slice, int> GetKnownFields()
+        {
+            if (_knownFields != null)
+                return _knownFields;
+            var knownFields = new Dictionary<Slice, int>();
+            foreach (var field in _fields.Values)
+            {
+                if(field.Name == null) 
+                    continue;
+                
+                Slice.From(_allocator, field.Name, ByteStringType.Immutable, out var value);
+                knownFields.Add(value, field.Id);
+            }
+
+            return knownFields;
+        }
+        
+        public Span<byte> InsertDocumentFields(LazyStringValue key, LazyStringValue sourceDocumentId, object doc, JsonOperationContext indexContext, out LazyStringValue id)
+        {
+            var document = (Document)doc;
+            _allocator.Allocate(document.Data.Size + _fields.Count * 1024, out ByteString buffer);
```

Comment:
Why `1024` ? 
When are you disposing this allocation?

Reply:
https://github.com/ravendb/ravendb/pull/12689/files#diff-93fa7c921f7580e39dabbfb8eef27076f75330cfb450cd0fc48f92814ba6ce35R82

Temporarily. I don't know how big a buffer needs to be to store all data per Document. This needs to be discussed. I know allocating buffer per document is a really bad idea for performance but nothing comes to my mind yet.

*marked as todo*